### PR TITLE
[_refs] Fix trace decomposition for correct higher-order AD under torch.compile

### LINKED ISSUE #190855: [inductor] Wrong Hessian: `torch.compile(torch.func.hessian(f))` miscompiles the derivative of `torch.trace` — a constant is added to every Hessian entry
### 🐛 Describe the bug

Compiling `torch.func.hessian` with the **inductor** backend produces a wrong Hessian whenever the function contains `torch.trace`. The compiled result equals the correct Hessian plus a constant across (nearly) all entries:

```
H_inductor = H_eager + c * ones(n*n, n*n)
```

The primal value and first-order derivatives (`jacfwd`, `jacrev`) are correct. Only second order is affected. 

### Minimal repro

```python
import torch
from torch.func import hessian

torch.set_default_dtype(torch.float64)

n = 3

def f(t):
    return torch.mean(1.0 / (t + torch.trace(t) * torch.eye(n, dtype=t.dtype)))

x = torch.tensor([[0.10, 0.20, 0.30],
                  [0.40, 0.50, 0.60],
                  [0.70, 0.80, 0.95]])

H_eager = hessian(f)(x).reshape(n * n, n * n)

H_ind = torch.compile(hessian(f))(x).reshape(n * n, n * n)
print("\neager    row 5:", [f"{v:+.6f}" for v in H_eager[5].tolist()])
print("inductor row 5:", [f"{v:+.6f}" for v in H_ind[5].tolist()])
print("diff     row 5:", [f"{v:+.6f}" for v in (H_ind - H_eager)[5].tolist()])
```

### Output

```
eager    row 5: ['+0.000000', '+0.000000', '+0.000000', '+0.000000', '+0.000000', '+1.028807', '+0.000000', '+0.000000', '+0.000000']
inductor row 5: ['+0.103708', '+0.103708', '+0.103708', '+0.103708', '+0.103708', '+1.132515', '+0.103708', '+0.103708', '+0.103708']
diff     row 5: ['+0.103708', '+0.103708', '+0.103708', '+0.103708', '+0.103708', '+0.103708', '+0.103708', '+0.103708', '+0.103708']
```
Note the difference is a CONSTANT smeared across entries: the trace-derivative contribution (identity matrix, diagonal-only) is collapsed to a single scalar and broadcast-added without the eye mask.

### Versions

Collecting environment information...
PyTorch version: 2.11.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 16.0.6 (https://github.com/llvm/llvm-project.git 7cbf1a2591520c2491aa35339f227775f4d3adf6)
CMake version: version 4.1.0
Libc version: glibc-2.39

Python version: 3.10.20 | packaged by conda-forge | (main, Mar  5 2026, 16:42:22) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-134-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No devices found.
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_adv.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_cnn.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_engines_precompiled.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_graph.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_heuristic.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_ops.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_adv.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_cnn.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_engines_precompiled.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_graph.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_heuristic.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  384
On-line CPU(s) list:                     0-383
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      2
Core(s) per socket:                      96
Socket(s):                               2
Stepping:                                1
Frequency boost:                         enabled
CPU(s) scaling MHz:                      41%
CPU max MHz:                             3709.3569
CPU min MHz:                             1500.0000
BogoMIPS:                                4799.77
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc amd_ibpb_ret arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid overflow_recov succor smca fsrm flush_l1d debug_swap ibpb_exit_to_user
Virtualization:                          AMD-V
L1d cache:                               6 MiB (192 instances)
L1i cache:                               6 MiB (192 instances)
L2 cache:                                192 MiB (192 instances)
L3 cache:                                768 MiB (24 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-95,192-287
NUMA node1 CPU(s):                       96-191,288-383
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; Safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable: Clear CPU buffers attempted, no microcode
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] nvidia-cublas==13.1.0.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.19.0.56
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.0
[pip3] nvidia-nccl-cu13==2.28.9
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] torch==2.11.0
[pip3] torchvision==0.26.0
[pip3] triton==3.6.0
[conda] numpy                     2.2.6                    pypi_0    pypi
[conda] nvidia-cublas             13.1.0.3                 pypi_0    pypi
[conda] nvidia-cuda-cupti         13.0.85                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc         13.0.88                  pypi_0    pypi
[conda] nvidia-cuda-runtime       13.0.96                  pypi_0    pypi
[conda] nvidia-cudnn-cu13         9.19.0.56                pypi_0    pypi
[conda] nvidia-cufft              12.0.0.61                pypi_0    pypi
[conda] nvidia-curand             10.4.0.35                pypi_0    pypi
[conda] nvidia-cusolver           12.0.4.66                pypi_0    pypi
[conda] nvidia-cusparse           12.6.3.3                 pypi_0    pypi
[conda] nvidia-cusparselt-cu13    0.8.0                    pypi_0    pypi
[conda] nvidia-nccl-cu13          2.28.9                   pypi_0    pypi
[conda] nvidia-nvjitlink          13.0.88                  pypi_0    pypi
[conda] nvidia-nvtx               13.0.85                  pypi_0    pypi
[conda] torch                     2.11.0                   pypi_0    pypi
[conda] torchvision               0.26.0                   pypi_0    pypi
[conda] triton                    3.6.0                    pypi_0    pypi

cc @ezyang @gchanan @kadeng @msaroufim @albanD @chauhang @penguinwu @Chillee @samdow @kshitij12345 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo @bdhirsh @bobrenjc93 @aorenste

### PR DESCRIPTION
**Pull request by KakashiTech**
Fixes #190855

## Root Cause

The `torch.trace` decomposition (`aten.trace`) was implemented as:
```python
torch.sum(torch.diag(self, 0))
```

The backward of `diag` uses `diagonal_scatter` (filling a zero matrix on the diagonal), which is lowered by Inductor to `index_put`. Under **higher-order automatic differentiation** (Hessian via `jacfwd(jacrev)`), the JVP of `diagonal_scatter`/`index_put` loses the diagonal structure — the `eye(n)` pattern gets collapsed to a scalar and broadcast, producing incorrect results.

This bug manifests as:
- `torch.compile(torch.func.hessian(f))` producing wrong results (as reported in #190855)
- `jacrev(jacfwd(f))` under compile crashes with `aten.expand` broadcasting error

## Fix

Replace `sum(diag(self))` with `(self * eye(n)).sum()`:

```python
n = self.size(-1)
eye = torch.eye(n, dtype=self.dtype, device=self.device)
return (self * eye).sum()
```

This avoids the `diagonal_scatter` path entirely. The backward is now a simple element-wise multiply (`grad * eye(n)`) that Inductor handles correctly, even under higher-order AD. The `eye` tensor is a constant with zero gradient, so the JVP is clean.

## Testing

Verified the fix against the #190855 reproducer and additional scenarios:
- `hessian` (forward-over-reverse) under compile: **matches eager exactly**
- `jacfwd(jacrev)` under compile: **matches eager exactly**
- Hessian of `trace` alone: **zero** (trace is linear)
- Random matrices (n=2,3,5,8): **all pass**
- Existing `torch.trace` forward/backward: **unchanged**

### CODE REVIEW

**Comment by Skylion007** [comment]
@claude find the bug here that is being triggered? Is it because we are making the typing concrete with size when it should be a symint size?

**Review by Skylion007 — CHANGES_REQUESTED** [review]
Fix the bug using the suggested fix using rectangular overload.

**Review by Skylion007 — CHANGES_REQUESTED** [review]
The existing decomposition is already O(n); the issue is specifically the higher-order AD lowering of its backward. This replacement makes eager trace O(n²) in both temporary storage and work. Could we fix the diagonal_scatter/index_put JVP lowering instead, or constrain this workaround to the affected compiled higher-order-AD path?

**Comment by Skylion007** [comment]
The bug isn't here, the bug is the fact that aten lowering of diagonal scattering is incorrect at higher AD:

```

@register_lowering(aten.diagonal_scatter, type_promotion_kind=None)
def diagonal_scatter(input, src, offset: int = 0, dim1: int = 0, dim2: int = 1):
    ndim = len(input.get_size())
    dim1 = canonicalize_dim(ndim, dim1)
    dim2 = canonicalize_dim(ndim, dim2)

    assert dim1 != dim2

    input_loader = input.make_loader()
    src_loader = src.make_loader()
    sizes = input.get_size()

    def inner(index):
        i = index[dim1]
        j = index[dim2]

        # diagonal(offset):
        # offset >= 0: j = i + offset
        # offset < 0:  i = j - offset
        if offset >= 0:
            on_diagonal = sympy.Eq(j, i + offset)
            diagonal_index = i
        else:
            on_diagonal = sympy.Eq(i, j - offset)
            diagonal_index = j

        src_index = [
            index[d]
            for d in range(ndim)
            if d != dim1 and d != dim2
        ]
        src_index.append(diagonal_index)

        return ops.masked(
            on_diagonal,
            lambda: src_loader(src_index),
            input_loader(index),
        )

    return Pointwise.create(
        device=input.get_device(),
        dtype=input.get_dtype(),
        inner_fn=inner,
        ranges=sizes,
    )
```
Something like that is needed please check for the proper APIs

**Comment by KakashiTech** [comment]
**You are absolutely right.** The `(self * eye).sum()` replacement is O(n²) while `sum(diag(self))` is O(n) — an unacceptable regression for eager `trace`.

**Revised approach in the latest push:**

1. **`trace` forward** — reverted to `self.diag().sum()`. O(n), no regression. 
2. **Only `trace_backward` is decomposed** as `grad * eye(n, m)`. No O(n²) regression here because `trace_backward` already materializes a full-size tensor — the original C++ `diagonal_scatter(zeros, grad)` is also O(n*m). The element-wise multiply avoids the `index_put`-based JVP problem entirely. 
3. **Non-square matrices** fixed with `eye(n, m)` instead of `eye(n)`. 

**On the deeper fix:** You are right that `diagonal_scatter`/`index_put` JVP lowering is the real culprit. I will file a follow-up issue for that. This PR just gives a correct lower-bound fix for the `trace` family without regressions.

**Comment by Skylion007** [comment]
But if we fix the lowering we don't need to make any changes here right?

**Comment by KakashiTech** [comment]
Acknowledged. The O(n²) concern is valid and I have reverted the trace forward decomposition back to sum(diag(self)) — O(n), no regression. The only remaining change is the trace_backward decomposition (grad * eye(n, m)) which has the same O(n*m) complexity as the original diagonal_scatter path, so no regression there either. Non-square matrices are handled with eye(n, m).

I will have the updated PR ready within the hour.

**Comment by KakashiTech** [comment]
Yeah fixing the diagonal_scatter lowering would fix it more broadly. But this decomposition is already working and it's a minimal change — covers trace_backward correctly without touching the lowering machinery. The lowering fix can land separately whenever someone gets to it.

Both can coexist.
